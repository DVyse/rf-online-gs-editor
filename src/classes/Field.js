/* eslint-disable default-case, no-fallthrough */
import { get } from 'lodash';

class Field {
  constructor({ type, name, len, as, schemaProps, writerProps }) {
    this.type = type;
    this.name = name;
    this.len = len;
    this.as = as;
    this.schemaProps = schemaProps || {};
    this.writerProps = writerProps || {};
    this.dynamic = typeof this.len === 'function';
  }

  isDynamic() {
    return this.dynamic;
  }

  getAs() {
    return this.as;
  }

  getName() {
    return this.name;
  }

  getType() {
    return this.type;
  }

  getLen(values = {}) {
    return typeof this.len === 'function' ? this.len(values) : this.len;
  }

  getWriterProps() {
    return this.writerProps;
  }

  getWriterProp(key, returnDefault = null) {
    return get(this.getWriterProps(), key, returnDefault);
  }

  getWeight(values = {}) {
    switch (this.getType()) {
      case Boolean:
      case Number:
        switch (this.getLen(values)) {
          case 8:
            return 1;
          case 16:
            return 2;
          case 32:
            return 4;
          case 64:
            return 8;
        }
      case String:
        switch (this.getAs()) {
          case 'hex':
            switch (this.getLen(values)) {
              case 8:
                return 1;
              case 16:
                return 2;
              case 32:
                return 4;
            }
          default:
            return this.getLen(values);
        }
    }

    throw new Error(`Undefined field type ${this.getType()}: ${this.getLen()}`);
  }

  toObject() {
    return {
      type: this.type,
      name: this.name,
    };
  }

  toSchema() {
    return {
      ...this.toObject(),
      props: {
        type: this.type,
        ...this.schemaProps,
      },
    };
  }
}

export default Field;
