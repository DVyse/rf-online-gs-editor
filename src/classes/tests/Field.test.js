import Field from '../Field';

describe('Classes', () => {
  describe('Field', () => {
    it('should throw invariant errors', () => {
      expect(() => new Field()).toThrow();
      expect(() => new Field({ type: Number, name: 's' })).toThrow();
      expect(
        () => new Field({ type: Number, name: 'valid', len: 'invalid' }),
      ).toThrow();

      expect(
        () =>
          new Field({ type: Number, name: 'valid', len: 32, as: 'invalid' }),
      ).toThrow();
    });
    it('should typeClassName defined', () => {
      expect(
        new Field({ type: Number, name: 'valid', len: 32 }).getTypeClassName(),
      ).toEqual('Number');

      expect(
        new Field({ type: Boolean, name: 'valid', len: 32 }).getTypeClassName(),
      ).toEqual('Boolean');

      expect(
        new Field({ type: String, name: 'valid', len: 32 }).getTypeClassName(),
      ).toEqual('String');
    });
    it('should typeClassName undefined', () => {
      const field = new Field({ type: String, name: 'valid', len: 32 });
      field.type = undefined;
      expect(field.getTypeClassName()).toBeUndefined();

      field.type = '__invalid_type__';
      expect(field.getTypeClassName()).toBeUndefined();

      field.type = { toString: () => '__invalid_type__' };
      expect(field.getTypeClassName()).toBeUndefined();
    });
    it('should return dynamic length', () => {
      const field = new Field({
        type: String,
        name: 'field',
        len: values => values.len,
      });

      expect(field.getLen({ len: 32 })).toEqual(32);
      expect(field.getLen({})).toEqual(undefined);
      expect(() => field.getLen()).toThrow();
      expect(new Field({ name: 'fl', type: Number, len: 32 }).getLen()).toEqual(
        32,
      );
    });
    it('should return Number|Boolean weights', () => {
      expect(
        new Field({ type: Number, len: () => 8, name: 'fl' }).getWeight({}),
      ).toEqual(1);

      expect(
        new Field({ type: Number, len: 8, name: 'fl' }).getWeight(),
      ).toEqual(1);

      expect(
        new Field({ type: Number, len: 16, name: 'fl' }).getWeight(),
      ).toEqual(2);

      expect(
        new Field({ type: Number, len: 32, name: 'fl' }).getWeight(),
      ).toEqual(4);

      expect(
        new Field({ type: Number, len: 64, name: 'fl' }).getWeight(),
      ).toEqual(8);

      expect(() =>
        new Field({ type: Number, len: 65, name: 'fl' }).getWeight(),
      ).toThrow();
    });
    it('should return String weights', () => {
      expect(
        new Field({ type: String, len: 8, name: 'fl', as: 'hex' }).getWeight(),
      ).toEqual(1);

      expect(
        new Field({ type: String, len: 16, name: 'fl', as: 'hex' }).getWeight(),
      ).toEqual(2);

      expect(
        new Field({ type: String, len: 32, name: 'fl', as: 'hex' }).getWeight(),
      ).toEqual(4);

      expect(() =>
        new Field({ type: String, len: 65, name: 'fl', as: 'hex' }).getWeight(),
      ).toThrow();

      expect(
        new Field({ type: String, len: 150, name: 'fl' }).getWeight(),
      ).toEqual(150);
    });
    it('should throw error on getWeight', () => {
      const field = new Field({ type: String, len: 32, name: 'fl', as: 'hex' });
      field.getType = () => '__unknown__';
      expect(() => field.getWeight()).toThrow();
    });
    it('should exist toObject', () => {
      expect(
        new Field({ type: Number, name: 'fl', len: 32 }).toObject(),
      ).toMatchSnapshot();
    });
    it('should exist toPlainObject', () => {
      expect(
        new Field({ type: Number, name: 'fl', len: 32 }).toPlainObject(),
      ).toMatchSnapshot();
      expect(
        new Field({ type: Number, name: 'fl', len: () => 32 }).toPlainObject(
          {},
        ),
      ).toBeDefined();
    });
    it('should exist toSchema', () => {
      expect(
        new Field({ type: Number, name: 'fl', len: 32 }).toSchema(),
      ).toMatchSnapshot();
    });
  });
});
