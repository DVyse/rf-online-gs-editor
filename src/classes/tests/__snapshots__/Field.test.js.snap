// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Classes Field should exist toObject 1`] = `
Object {
  "name": "fl",
  "type": [Function],
}
`;

exports[`Classes Field should exist toPlainObject 1`] = `
Object {
  "as": undefined,
  "len": 32,
  "name": "fl",
  "schemaProps": Object {},
  "type": [Function],
  "typeClassName": "Number",
  "weight": 4,
}
`;

exports[`Classes Field should exist toSchema 1`] = `
Object {
  "name": "fl",
  "props": Object {
    "type": [Function],
  },
  "type": [Function],
}
`;
